### DISCUSSION TITLE
Scale Ticks do not update when data is updated (filtered)

### DISCUSSION THREAD

**Discussion by Kaspatoo** [kurkle/chartjs-chart-matrix#70] [Q&A]
Hi,
in a normal JSChart the scale ticks also update when the underlying data is updated.

E.g. there is on x-axis the ticks `{a,b,c}` and data for all of them.
After filtering there is no more data for `'b'`. While the matrix-tile disappears the 'b'-tick still stays in place.
In normal ChartJS 'b' would also disappear now. In Matrix it doesnt.

In Pictures:
How it looks at the beginning:
![image](https://user-images.githubusercontent.com/17131190/186610333-27a0becc-4596-47e8-934c-a05251a2bd1d.png)

How it should look after filtering data (in this case no more data for [B,Z]):
![image](https://user-images.githubusercontent.com/17131190/186610694-aafbce90-08eb-4e69-a071-f04029cd1074.png)

How it actually looks:
![image](https://user-images.githubusercontent.com/17131190/186610630-84568feb-5509-4919-93bc-b4e5618840ec.png)

I tried several play arounds with the labels-array of a `dataset`' but it seems that chartjs-matrix is ignoring this array.
Instead I was able to achieve tiny changes when working on `scales/x/labels`.
When using an empty array `[]` the chartjs-matrix gathers the labels on its own by the given intial data (in this sample this is `{a,b,c}`).
This does not change after updating the data.
When using a non-empty array `['a','b','c']}` this leads to the same behavior.

`scales: {
                              x: {
                                  title: {
                                    display: true,
                                    text: 'Grade'
                                  },
                                  position: 'top',
                                  type: 'category',
                                  labels: ['d','e','f'],
                                  ticks: {
                                      display: true
                                  },
                                  grid: {
                                      display: false
                                  }
                              },`

I tried overwriting any array in the dataset or the chart, but did not have luck yet.
There are severy places knowing of scale-ticks possibly reliable. A short print-out:
![image](https://user-images.githubusercontent.com/17131190/186613500-72379851-99e3-4574-a0dc-ca51d0527406.png)

Any ideas?

**Comment by Kaspatoo** [comment]
as a workaround I also failed with that:
(we extended react.component)

`

    componentDidMount() {
          const config = {
            type: this.props.type,
            plugins: [ChartDataLabels],
            data: this.state.data,
            options: this.props.options ? this.props.options : {}
          };
          this.chart = new Chart(document.getElementById(this.props.id),  config);
    }

    componentDidUpdate() {
        this.chart.destroy();
        this.componentDidMount();
        this.chart.data = this.state.data;
        this.chart.update();
    }`

Its just the same behavior.

**Comment by kurkle** [comment]
There are several places where the failure could be (the scale, the component and (maybe most likely) matrix chart). Could you create a very simple codepen to reproduce?

**Comment by Kaspatoo** [comment]
I dont know how
is there a codepen available from the samples? maybe one with the functionality to update?

**Comment by kurkle** [comment]
You can fork that:
https://codepen.io/kurkle/pen/qBbeOZN

And I guess you can do the update right after the chart is created.

**Comment by Kaspatoo** [comment]
can you extend it having a button which calls the update chart function giving new data into?

**Comment by Kaspatoo** [comment]
there it is:
https://codepen.io/kaspatoo/pen/zYWQPoE

i guess its having to do with setting the data directly but not using pop on the dataset? may some eventlistener is thrown away?
but as said for normal chartjs charts its working this way.

**Comment by Kaspatoo** [comment]
@kurkle have you noticed my code sample? didnt hear something

**Comment by Kaspatoo** [comment]
popping and pushing elements leads to the same effect:

`    componentDidUpdate() {
        var origin = this.chart.data.datasets;
        var destination = this.state.data.datasets;

        //Intially set data
        if(origin.length == 0){
            this.chart.data = this.state.data;
        }else{
            var origin_data = origin[0].data;
            var destination_data = destination[0].data;

            //clear origin first
            while(origin_data.length > 0){
                origin_data.pop();
            }
            //then fill again
            destination_data.forEach(item => { origin_data.push(item);});
        }
        this.chart.update();
    }`

**Comment by kurkle** [comment]
Sorry for the late reply, but this is the `cagegory` scale caching its labels. You can work around it by clearing the labels array in scale options before update:

```js
  chart.data.datasets[0].data = local_data2;
  // these next 2 lines
  chart.scales.x.options.labels = [];
  chart.scales.y.options.labels = [];
  chart.update();
```

**Reply by Kaspatoo** [reply]
perfect, nice
couldnt you overwrite the update-function and include that? I mean it only happens in matrix chart so...

**Reply by Kaspatoo** [reply]
(just as future feature, for me you solution is great enough)

**Reply by kurkle** [reply]
> perfect, nice
> 
> couldnt you overwrite the update-function and include that? I mean it only happens in matrix chart so...

Intereating, I have to re-check what is done in the core and does this only happen when both scales are category.
